src/Fantomas/CodeFormatterImpl.fs

```diff
@@ -44,6 +44,8 @@ let createFormatContext fileName source projectOptions checker =
 
 let parse { FileName = fileName; Source = source; ProjectOptions = checkOptions; Checker = checker } = 
     async {
+        // This defines are needed for the Tokenizer
+        TokenMatcher.CompilationDefines <- checkOptions.ConditionalCompilationDefines
```

Comment:
This seems a bit strange, what if we are parsing multiple files? Will the `CompilationDefines` be reset for each file or will they contains everything needed for all files?

Should we not return another type from the [parse](https://github.com/fsprojects/fantomas/blob/master/src/Fantomas/CodeFormatterImpl.fs#L45) functions that has the `ParsedInput` and the defines?

Reply:
``ConditionalCompilationDefines`` are the same for a whole Project, F# doesn't support ``#define``, does it? So formatting of multiple files in parallel is not a problem. A problem were to format multi projects in parallel (threads) but is it possible at all?

That said, i have nothing vs a thread safe solution but as far I understand it changing ``parse`` signature would require public API changes. I am not sure what sort of consensus needed here.

Reply:
Multiple projects does not work today, but is requested in https://github.com/fsprojects/fantomas/issues/251. I guess it could be a problem some day.

`parse` is indeed public api, maybe [format](https://github.com/fsprojects/fantomas/blob/master/src/Fantomas/CodeFormatterImpl.fs#L369) isn't? What if we get the info needed inside `parse` and collect it in `format` instead and pass it to `formatWith` down the line.

Reply:
but ``formatAST`` is a part of public API and would need a signature change. I could wrap (change signature) ``parse`` like you suggested previously and unwrap in the API, that would work as long no one use ``parse`` directly outside there. 

The cleanest approach seems to me to change API but i would say #251 is better place for a API change as probably more things are needed than ``Defines`` forwarding.

Reply:
I'm a bit in the dark if changing the API is really a bad thing. The only editor I know that is using a recent version of Fantomas is Rider, perhaps @auduchinok could tell us if the change would have a lot of impact.

Most people are using Fantomas from the command line, or that is my assumption.
@dsyme do you have any feedback?

Reply:
We're fine with changing signature of `FormatAST` (or anything else in the API) as long as it's possible to fix our use for the change so it works the same way.

Reply:
I've removed the global variable and changed API as needed.